src/coreclr/jit/perfscorearm64sven2.h

```diff
@@ -0,0 +1,868 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+/*****************************************************************************
+* Perf scores for Arm N2 SVE instructions
+* Taken from Arm Neoverse N2 Software Optimization Guide, Issue 5.0, Revision: r0p3
+*****************************************************************************/
+
+// CRC checksum ops
+#define PERFSCORE_LATENCY_N2_CRC_CHECKSUM_OPS PERFSCORE_LATENCY_2C
+#define PERFSCORE_THROUGHPUT_N2_CRC_CHECKSUM_OPS PERFSCORE_THROUGHPUT_1C
+
+// Loop control, based on predicate
+#define PERFSCORE_LATENCY_N2_LOOP_CONTROL_BASED_ON_PREDICATE PERFSCORE_LATENCY_2C
+#define PERFSCORE_THROUGHPUT_N2_LOOP_CONTROL_BASED_ON_PREDICATE PERFSCORE_THROUGHPUT_2X
+
+// Loop control, based on predicate and flag setting
+#define PERFSCORE_LATENCY_N2_LOOP_CONTROL_BASED_ON_PREDICATE_AND_FLAG_SETTING PERFSCORE_LATENCY_3C
+#define PERFSCORE_THROUGHPUT_N2_LOOP_CONTROL_BASED_ON_PREDICATE_AND_FLAG_SETTING PERFSCORE_THROUGHPUT_2X
+
+// Loop control, propagating
+#define PERFSCORE_LATENCY_N2_LOOP_CONTROL_PROPAGATING PERFSCORE_LATENCY_2C
+#define PERFSCORE_THROUGHPUT_N2_LOOP_CONTROL_PROPAGATING PERFSCORE_THROUGHPUT_1C
+
+// Loop control, propagating and flag setting
+#define PERFSCORE_LATENCY_N2_LOOP_CONTROL_PROPAGATING_AND_FLAG_SETTING PERFSCORE_LATENCY_3C
```

Comment:
> Are they different from what is defined in [emitPerfScore_sve.cpp](https://github.com/dotnet/runtime/files/13502454/emitPerfScore_sve.cpp.txt) that I have uploaded in #94549?

I hadn't spotted the latest update of `emitPerfScore_sve.cpp`.

But, I now realise where the errors are coming in .... for PERFSCORE_THROUGHPUT you've  mixed up the X and C cases.  X is when there are more than one per cycle. Looking at the file, INS_sve_fsqrt is marked as `PERFSCORE_THROUGHPUT_14X` which can't be right as it's always going to be a slow instruction.

With that fixed I'm thinking maybe we should stick to your file instead of this PR.

What is missing from your file is the cases where the perfscore changes depending on the element size (eg `IF_SVE_AK_3A`). I noticed that NEON has these.

I also think for currently unsupported instructions we should set the score high, rather than 1C.


> They are purely used for display purpose and we do not depend on them for any optimizations.

What are they used for? Is it just for manually comparing the relative performance of one version of a compiled routine against another? Deciding whether a new coreclr optimisation improves or regresses the code?

If so, then my concern is that the relative performance of difference instructions can change quite a bit over time depending on microarchitecture design, and especially between different types of CPUs (eg: a mobile chip vs a server one) as they are design for different purposes. So, basing decisions on old architectures won't match the actual performance seen on real hardware. Could quite easily add an optimisation that that has a better perfscore, but worse real performance. Right now, comparing a NEON loop against an SVE loop using the existing perfscore data would be a bad idea, as I suspect NEON comes from a much older chip.

Ok, I can see why you wouldn't want multiple sets of perfscores. But, on the other hand I don't think a mix up of scores from a variety of sources is helpful. I'm not sure what a good solution would be though.

Reply:
Perfscore is currently an estimate of performance that can be used in asmdiffs to determine if a change is good from a performance perspective, not just a code size perspective. See "superpmi.py asmdiffs -metrics" and jit-analyze "--metrics".

We're not sure how good these scores are. @AndyAyersMS has been doing various analyses to determine how usable perfscore is for (auto) tuning CSE instead of running benchmarks and measuring time (which is noisy, etc.).

I think choosing a particular platform and using it for all the data makes sense. Whether we ever change the set of numbers we use based on the particular target platform is TBD. We're nowhere close to being able to make that decision.  Updating the existing numbers to be based on a particular modern NEON platform makes sense, but might not be worth the effort, either.

Reply:
> X is when there are more than one per cycle. Looking at the file, INS_sve_fsqrt is marked as `PERFSCORE_THROUGHPUT_14X` which can't be right as it's always going to be a slow instruction.

The way `X` is currently used for `throughput < 1`. `fsqrt`, that sounds correct because it is `1/14` as minimum. 

<img width="601" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/f963c209-211b-4fb9-9c12-354337369ad5">

> What are they used for? Is it just for manually comparing the relative performance of one version of a compiled routine against another? Deciding whether a new coreclr optimisation improves or regresses the code?

That's right. We do not even have them on by default in `spmi-diffs` pipeline, the default we use is `codeSize`. But for register allocation changes, `PerfScore` comes handy because those changes can increase the code size but eliminate memory access and that increases the code size, but lowers the perfscore.

> If so, then my concern is that the relative performance of difference instructions can change quite a bit over time depending on microarchitecture design, and especially between different types of CPUs 

Yes, that is a true statement and we do not keep it upto date.

> Right now, comparing a NEON loop against an SVE loop using the existing perfscore data would be a bad idea, as I suspect NEON comes from a much older chip.

I don't think we would do that. We will just compare the disassembly :)

> So, basing decisions on old architectures won't match the actual performance seen on real hardware

pretty much to what @BruceForstall mentioned about our experimentation in this area.

Reply:
> Right now, comparing a NEON loop against an SVE loop using the existing perfscore data would be a bad idea, as I suspect NEON comes from a much older chip.

Would it make sense to just update all the data to be from a single chip then?